plugins/user.py

```diff
@@ -0,0 +1,220 @@
+            self.say(msg)
+        self.TWOFACTOR_S3_BUCKET = settings.TWOFACTOR_S3_BUCKET
+        if not hasattr(settings, "TWOFACTOR_S3_PROFILE"):
+            msg = "Error: TWOFACTOR_S3_PROFILE not defined in the environment"
+            self._say_error(msg)
+        self.TWOFACTOR_S3_PROFILE = settings.TWOFACTOR_S3_PROFILE
+
+
+            
+
+
+    def generate_and_upload_QR(self,secret,username):
+        try:
+            principle = self.TWOFACTOR_PRINCIPLE
+            issuer = self.TWOFACTOR_ISSUER
+            s3_twofactor_bucket = self.TWOFACTOR_S3_BUCKET
+            s3_twofactor_profile = self.TWOFACTOR_S3_PROFILE
+        except:
+            return "Settings for S3 hosted QR codes not configured properly in environment"
+
+        url = 'otpauth://totp/{issuer}:{principle}?secret={secret}&issuer={issuer}'.format(issuer=issuer, principle=principle, secret=secret)
+        img = qrcode.make(url)
+        conn = boto.connect_s3(profile_name=s3_twofactor_profile)
+        bucket = conn.get_bucket(s3_twofactor_bucket)
+        key = Key(bucket)
+        key.key = '{}-qr.png'.format(username)
+        image_buffer = cStringIO.StringIO()
+        img.save(image_buffer, 'PNG')
+        key.set_contents_from_string(image_buffer.getvalue())
+        url = key.generate_url(60,query_auth=True, force_http=True)
+        return url
+
+    @respond_to("^twofactor verify (?P<token>\w*)")
+    def verify_user_twofactor(self, message, token):
+        """
+        twofactor verify [token]: verify your twofactor authentication system
+        """
+        username = message.sender.nick
+        if verify_twofactor(self,username, token):
+            self.reply(message, "You are authenticated, {}".format(username)) 
+        else:
+            self.reply(message, "That's not correct,{}".format(username))
+
+
+    @respond_to("^twofactor me")
+    def create_user_twofactor(self, message):
+        """
+        twofactor me: set up twofactor authentication for your user
+        """
+        username = message.sender.nick
+        user_twofactor = self.load("user_twofactor", {})
+        if username not in user_twofactor:
+            user_twofactor[username] = pyotp.random_base32()
+            message['type'] = 'chat'
+            self.reply(message, self.generate_and_upload_QR(user_twofactor[username], username))
+            self.save("user_twofactor", user_twofactor)
+            self.reply(message, "say 'twofactor verify <token>' to check your twofactor verification")
+        else:
+            message['type'] = 'chat'
+            self.reply(message, "twofactor is already set up!") 
+
+    @respond_to("^twofactor remove (?P<username>\w*)")
+    def remove_user_twofactor(self, message, username):
+        """
+        twofactor remove [username]: remove [username]'s twofactor authentication (requires 'admin' permission)
+        """
+        if check_user_permission(self, message.sender.nick, 'admin'):  
```

Comment:
Might be better to bind permissions to a unique user identifier rather than the nick (which can change). Also, check out [bouncer](https://pypi.python.org/pypi/bouncer/0.1.10)!